pkg/pubsubsource/pubsub_test.go

```diff
@@ -0,0 +1,221 @@
+			if strings.Contains(name, "google-cloud-cli") {
+				pubSubRunning = true
+				break
+			}
+		}
+		if pubSubRunning {
+			break
+		}
+	}
+
+	if !pubSubRunning {
+		// Start goaws container if not already running
+		opts := dockertest.RunOptions{
+			Repository:   "thekevjames/gcloud-pubsub-emulator",
+			Tag:          "latest",
+			ExposedPorts: []string{"8681"},
+			PortBindings: map[docker.Port][]docker.PortBinding{
+				"8681": {
+					{HostIP: "127.0.0.1", HostPort: "8681"},
+				},
+			},
+		}
+		resource, err = pool.RunWithOptions(&opts)
+		if err != nil {
+			_ = pool.Purge(resource)
+			log.Fatalf("could not start resource %s", err)
+		}
+	}
+
+	err = os.Setenv("PUBSUB_EMULATOR_HOST", "localhost:8681")
+	if err != nil {
+		log.Fatalf("error -%s", err)
+	}
+	ctx := context.Background()
+
+	if err := pool.Retry(func() error {
+		// Creates a client.
+		pubsubClient, err = pubsub.NewClient(ctx, Project)
+		if err != nil {
+			log.Fatalf("Failed to create client: %v", err)
+		}
+		return nil
+	}); err != nil {
+		if resource != nil {
+			_ = pool.Purge(resource)
+		}
+		log.Fatalf("could not connect to gcloud pubsub sqs %s", err)
+	}
+	code := m.Run()
+	if resource != nil {
+		if err := pool.Purge(resource); err != nil {
+			log.Fatalf("Couln't purge resource %s", err)
+		}
+	}
+	defer pubsubClient.Close()
+	os.Exit(code)
+}
+
+func TestPubSubSource_Read(t *testing.T) {
+
+	err := ensureTopicAndSubscription(context.TODO(), pubsubClient, TopicID, subscriptionID)
+	assert.Nil(t, err)
+	messageCh := make(chan sourcer.Message, 20)
+	//doneCh := make(chan struct{})
+	sub := pubsubClient.Subscription(subscriptionID)
+	pubsubsource := NewPubSubSource(pubsubClient, sub)
+	ctx := context.Background()
+	go func() {
+		<-time.After(5 * time.Second)
+		sendMessages(context.Background(), pubsubClient, TopicID, 5)
+	}()
+
+	pubsubsource.Read(ctx, mocks.ReadRequest{
+		CountValue: 5,
+		Timeout:    20 * time.Second,
+	}, messageCh)
+	assert.Equal(t, 5, len(messageCh))
+
+	// Try reading 4 more messages
+	// Since the previous batch didn't get acked, the data source shouldn't allow us to read more messages
+	// We should get 0 messages, meaning the channel only holds the previous 5 messages
+	// Creating a new subscriber
+	pubsubsource.Read(ctx, mocks.ReadRequest{
+		CountValue: 4,
+		Timeout:    10 * time.Second,
+	}, messageCh)
+	assert.Equal(t, 5, len(messageCh))
+
+	// Ack the first batch
+	// Acknowledge messages.
+	for i := 0; i < 5; i++ {
+		msg := <-messageCh
+		pubsubsource.Ack(ctx, mocks.TestAckRequest{
+			OffsetsValue: []sourcer.Offset{msg.Offset()},
+		})
+	}
+	go func() {
+		<-time.After(5 * time.Second)
+		sendMessages(context.Background(), pubsubClient, TopicID, 6)
+
+	}()
+
+	// Try reading 6 more messages
+	// Since the previous batch got acked, the data source should allow us to read more messages
+	// We should get 6 messages
+	sub2 := pubsubClient.Subscription(subscriptionID)
+	pubsubsource2 := NewPubSubSource(pubsubClient, sub2)
```

Comment:
why do we create a new source? The `read` and `ack` are co-ordinating within a single source context. If we create a separate source 2, would it always succeed in reading the 6 messages no matter whether source 1 ack or not?

Reply:
Removed ,now creating a new subscriber ,as we can create multiple subscribers from same subscription .I am closing the previous subscriber as its not usable post the Read operation is finished